sympy/core/exprtools.py

```diff
@@ -245,47 +246,50 @@ def decompose_power(expr):
     if exp.is_Number:
         if exp.is_Rational:
             if not exp.is_Integer:
-                base = Pow(base, Rational(1, exp.q))
-
-            exp = exp.p
+                base = Pow(base, Rational(1, cast(Rational, exp).q))
+            e = cast(Rational, exp).p
```

Comment:
I wonder how we would have to go with adding `cast` like this in different parts of the codebase. Is there a way somehow to get mypy to understand that for `Basic` instances `exp.is_Rational` is equivalent to `isinstance(exp, Rational)`?

Reply:
```py
import functools

class A:
    is_B = False
    a = 0

    @functools.cached_property
    def method(self):
        return isinstance(self, B)

class B(A):
    is_B = True
    a = 1
    b = 2

def f(a: A):
    print(a.a)
    if isinstance(a, B):
        print(a.b)
    if a.is_B:
        print(a.b)
    if a.method:
        print(a.b)

f(B())
```
```
hint.py:21: error: "A" has no attribute "b"
hint.py:23: error: "A" has no attribute "b"
```
This minimal example shows even using property mypy cannot tell a is B. I expect @overload could be used to decorate a field like is_Rational in future python version. Now `cast` has to be used to avoid `# type: ignore`.

Reply:
I think I'd rather have a `# type: ignore` than the cast. Adding hints in function signatures or variable declarations to "annotate" the code is one thing but complicating the actual code with extra bits that don't do anything seems like too much to me.

Besides making improvements to mypy and typing to handle this case I can see several options here:

1. Change `exp.is_Rational` to `isinstance(exp, Rational)`. In principle I would rather write the code like this anyway but it is actually slower at runtime and also exacerbates circular import problems.
2. Use `type: ignore`. This is considered undesirable if you actually want the checker to check things for you because it effectively disables checking. Since it's only changing the code by adding a comment though this doesn't mess up the main code.
3. Use `cast`. This has a runtime cost *and* messes up the code *and* also causes problems with circular imports. Effectively the use of `cast` also disables checking at least the point is to override the checker so what is stated by the case is not actually checked.

I'm not sure why `cast` would be considered the preferred option here. Really I think it would be good to go ask the mypy people how this can be done or suggest that mypy be improved to handle this case rather than compromising on other things just for the sake of the checker.

Reply:
https://docs.python.org/3/library/typing.html#typing.cast cast shouldn't have runtime cost. I use that because `# type: ignore` ignores the whole line which may not be what we want. But I'm OK to update this PR to use `# type: ignore` instead.

Reply:
Calling any function in Python has a runtime overhead even if the function doesn't do anything:
```python
In [6]: from typing import cast

In [7]: x = Symbol('x')

In [8]: timeit x.is_Rational
81.5 ns ± 4.45 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [9]: timeit isinstance(x, Rational)
330 ns ± 3.1 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [10]: timeit cast(x, Rational)
225 ns ± 6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
More importantly though the cast obfuscates what should be a very simple line of code making it harder to read.